sklearn/cluster/optics_.py

```diff
@@ -0,0 +1,774 @@
+            self.reachability_[unproc] = new_reach
+
+        # Checks to see if everything is already processed;
+        # if so, return control to main loop
+        if unproc.size > 0:
+            # Define return order based on reachability distance
+            return(unproc[quick_scan(np.take(self.reachability_, unproc),
+                                     dists)])
+        else:
+            return point_index
+
+    def extract_dbscan(self, eps):
+        """Performs DBSCAN extraction for an arbitrary epsilon.
+
+        Extraction runs in linear time. Note that if the `max_bound` OPTICS
+        parameter was set to < inf for extracting reachability and ordering
+        arrays, DBSCAN extractions will be unstable for `eps` values close to
+        `max_bound`. Setting `eps` < (`max_bound` / 5.0) will guarantee
+        extraction parity with DBSCAN.
+
+        Parameters
+        ----------
+        eps : float or int, required
+            DBSCAN `eps` parameter. Must be set to < `max_bound`. Equivalence
+            with DBSCAN algorithm is achieved if `eps` is < (`max_bound` / 5)
+
+        Returns
+        -------
+        core_sample_indices_ : array, shape (n_core_samples,)
+            The indices of the core samples.
+
+        labels_ : array, shape (n_samples,)
+            The estimated labels.
+        """
+        check_is_fitted(self, 'reachability_')
+
+        if eps > self.max_bound:
+            raise ValueError('Specify an epsilon smaller than %s. Got %s.'
+                             % (self.max_bound, eps))
+
+        if eps * 5.0 > (self.max_bound * 1.05):
+            warnings.warn(
+                "Warning, max_bound (%s) is close to eps (%s): "
+                "Output may be unstable." % (self.max_bound, eps),
+                RuntimeWarning, stacklevel=2)
+        # Stability warning is documented in _extract_dbscan method...
+
+        return _extract_dbscan(self.ordering_, self.core_distances_,
+                               self.reachability_, eps)
+
+
+def _extract_dbscan(ordering, core_distances, reachability, eps):
+    """Performs DBSCAN extraction for an arbitrary epsilon (`eps`).
+
+    Parameters
+    ----------
+    ordering : array, shape (n_samples,)
+        OPTICS ordered point indices (`ordering_`)
+    core_distances : array, shape (n_samples,)
+        Distances at which points become core (`core_distances_`)
+    reachability : array, shape (n_samples,)
+        Reachability distances calculated by OPTICS (`reachability_`)
+    eps : float or int
+        DBSCAN `eps` parameter
+
+    Returns
+    -------
+    core_sample_indices_ : array, shape (n_core_samples,)
+        The indices of the core samples.
+
+    labels_ : array, shape (n_samples,)
+        The estimated labels.
+    """
+
+    n_samples = len(core_distances)
+    is_core = np.zeros(n_samples, dtype=bool)
+    labels = np.ones(n_samples, dtype=int) * -1
+    cluster_id = 0
+
+    for entry in ordering:
```

Comment:
You can do the is_core determination en mass, but I'm not sure about assigning labels... I think the run time on this is pretty negligible, since it's just a single linear scan.

Reply:
Generally if there's a python loop over n_samples that we can avoid, we try to. I also think the numpy logic is a more succinct way of expressing the various cases. But I've not checked that my code matches yours.